framework/src/variables/MooseVariableDataBase.C

```diff
@@ -446,22 +439,15 @@ MooseVariableDataBase<RealEigenVector>::add(NumericVector<Number> & residual)
 {
   if (_has_dof_values)
   {
-    auto & dof_values = _vector_tags_dof_u[_solution_tag];
-    if (isNodal())
+    const auto & dof_values = _vector_tags_dof_u[_solution_tag];
+    mooseAssert(_dof_indices.size() % _count == 0,
+                "Dof indices should be cleanly divisible by the variable count");
+    const auto n_shapes = _dof_indices.size() / _count;
+    for (const auto indx : index_range(_dof_indices))
```

Comment:
we're losing the potential to vectorize over the vector in the nodal case. we had a clear access pattern and now we're accessing inside a vector to get the index

Reply:
The thing is that the old code had zero robustness. Even for a `isNodal()` `LAGRANGE_VEC` variable this would not have worked because the old `_dof_indices` behavior assumed a stride of 1 between variable components in the global dof index numbering but with a `LAGRANGE_VEC` that stride is no longer 1. We may lose some performance here but I much prefer robustness over speed

Reply:
the stride between the same shape function for different array variable components is the number of shapes for a single component, e.g. `n_shapes`